level_5/parser.hpp

```diff
@@ -0,0 +1,190 @@
+};
+
+/// <Operand> Operator <Operand> Binary operator class
+/// with two (left and right) operands
+class BinOp final : public ASTNode {
+      public:
+        BinOp(std::unique_ptr<ASTNode> left, std::unique_ptr<ASTNode> right,
+              const Token& token)
+            : left_node_{std::move(left)},
+              right_node_{std::move(right)}, token_{token} {}
+        const Token& get_token() const noexcept override { return token_; }
+        const NodeType get_node_type() const noexcept override {
+                return NodeType::BinOp;
+        }
+        ASTNode* get_left() noexcept override { return left_node_.get(); }
+        ASTNode* get_right() noexcept override { return right_node_.get(); }
+        static std::unique_ptr<ASTNode> make(std::unique_ptr<ASTNode> left,
+                                             std::unique_ptr<ASTNode> right,
+                                             const Token& token) {
+                return std::make_unique<BinOp>(std::move(left),
+                                               std::move(right), token);
+        }
+
+      private:
+        Token token_;
+        std::unique_ptr<ASTNode> left_node_{nullptr};
+        std::unique_ptr<ASTNode> right_node_{nullptr};
+};
+
+/// <Operator> Operand Unary operator class
+/// with one operand
+class UnaryOp final : public ASTNode {
+      public:
+        UnaryOp(std::unique_ptr<ASTNode> left, const Token& token)
+            : token_{token}, left_node_{std::move(left)} {}
+        const Token& get_token() const noexcept override { return token_; }
+        const NodeType get_node_type() const noexcept override {
+                return NodeType::UnaryOp;
+        }
+        ASTNode* get_left() noexcept override { return left_node_.get(); }
+        ASTNode* get_right() noexcept override { return right_node_.get(); }
+        static std::unique_ptr<ASTNode> make(std::unique_ptr<ASTNode> left,
+                                             std::unique_ptr<ASTNode>,
+                                             const Token& token) {
+                return std::make_unique<UnaryOp>(std::move(left), token);
+        }
+
+      private:
+        Token token_;
+        std::unique_ptr<ASTNode> left_node_{nullptr};
+        std::unique_ptr<ASTNode> right_node_{nullptr};
+};
+
+class Parser {
+      public:
+        /// Interpreter constructor for the calculator
+        ///
+        /// \param lexer Lexer for getting the tokens, one at a time
+        explicit Parser(const Lexer& lexer)
+            : lexer_(lexer), current_token_(lexer_.get_next_token()) {}
+        /// Let's Lex the input expression, parse it and build an AST
+        ///
+        /// \return Handler to the root node of the AST
+        std::unique_ptr<ASTNode> parse() { return std::move(expr()); }
+
+      private:
+        Lexer lexer_;
+        Token current_token_;
+        void error() { throw std::logic_error("Invalid syntax"); }
```

Comment:
I was already expressing my feelings about "throwing", but here in this case, at least name the function appropriately, what about: `raiseInvalidSyntaxError()`?